framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,876 @@
+package org.checkerframework.framework.util;
+
+import java.lang.annotation.Annotation;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.checker.interning.qual.Interned;
+import org.checkerframework.framework.qual.PolymorphicQualifier;
+import org.checkerframework.framework.qual.SubtypeOf;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.SystemUtil;
+import org.checkerframework.javacutil.UserError;
+
+/**
+ * This class holds information about the subtyping relationships between qualifiers.
+ *
+ * <p>By default, the subtyping information and information about polymorphic qualifiers is read
+ * from meta-annotations on the annotation classes. This information is used to infer further
+ * information such as top and bottom qualifiers. Subclasses can override the following methods to
+ * change this behavior:
+ *
+ * <ul>
+ *   <li>{@link #createQualifierKinds(Collection)}
+ *   <li>{@link #createDirectSuperMap()}
+ *   <li>{@link #specifyBottom(Map, Class)}
+ *   <li>{@link #initializePolymorphicQualifiers()}
+ *   <li>{@link #initializeQualifierKindFields(Map)}
+ *   <li>{@link #createLubsMap()}
+ *   <li>{@link #createGlbsMap()}
+ * </ul>
+ *
+ * This class is used by {@link SimpleQualifierHierarchy} and {@link ComplexQualifierHierarchy}
+ * classes to implement methods that compare {@link javax.lang.model.element.AnnotationMirror}s,
+ * such as {@link org.checkerframework.framework.type.QualifierHierarchy#isSubtype(AnnotationMirror,
+ * AnnotationMirror)}.
+ */
+public class QualifierKindHierarchy {
+
+    /**
+     * Represents a qualifier class. It holds information about the relationship between itself and
+     * other {@link QualifierKind}s.
+     *
+     * <p>Exactly one qualifier kind is created for each annotation class.
+     */
+    // The private non-final fields of this class are set while creating the QualifierKindHierarchy.
+    public static @Interned class QualifierKind implements Comparable<QualifierKind> {
```

Comment:
I had originally named this class "QualifierClass", but it is confusing because a qualifier class would have type `Class<Annotation>`. For example, see the constructor of `QualifierKindHierarchy`.

This class does not represent a "Qualifier" because a qualifier may have values for the annotation elements, but QualifierKinds do not.  So `@Regex(1)` is a qualifier but it can't be represents by this class.

Reply:
TODO: Document why it's called kind.